app/grandchallenge/algorithms/forms.py

```diff
@@ -365,20 +366,87 @@ class Meta:
             "workstation_config": "Viewer Configuration",
         }
 
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
+    def __init__(self, *args, user, **kwargs):
+        super().__init__(*args, user=user, **kwargs)
+        self._user = user
 
         self.fields["contact_email"].required = True
         self.fields["display_editors"].required = True
 
-        self.fields["job_requires_gpu_type"].choices = [
-            (choice.value, choice.label)
-            for choice in [GPUTypeChoices.NO_GPU, GPUTypeChoices.T4]
-        ]
+        self.fields["job_requires_gpu_type"].choices = (
+            self.selectable_gpu_type_choices
+        )
         self.fields["job_requires_memory_gb"].validators = [
             MinValueValidator(settings.ALGORITHMS_MIN_MEMORY_GB),
-            MaxValueValidator(settings.ALGORITHMS_MAX_MEMORY_GB),
+            MaxValueValidator(self.maximum_settable_memory_gb),
+        ]
+
+    @cached_property
+    def job_requirement_properties_from_phases(self):
+        qs = get_objects_for_user(
+            self._user, "evaluation.create_phase_submission"
+        )
+        inputs = self.instance.inputs.all()
+        outputs = self.instance.outputs.all()
+        return (
+            qs.annotate(
+                total_algorithm_input_count=Count(
+                    "algorithm_inputs", distinct=True
+                ),
+                total_algorithm_output_count=Count(
+                    "algorithm_outputs", distinct=True
+                ),
+                relevant_algorithm_input_count=Count(
+                    "algorithm_inputs",
+                    filter=Q(algorithm_inputs__in=inputs),
+                    distinct=True,
+                ),
+                relevant_algorithm_output_count=Count(
+                    "algorithm_outputs",
+                    filter=Q(algorithm_outputs__in=outputs),
+                    distinct=True,
+                ),
+            )
+            .filter(
+                submission_kind=SubmissionKindChoices.ALGORITHM,
+                public=True,
+                total_algorithm_input_count=len(inputs),
+                total_algorithm_output_count=len(outputs),
+                relevant_algorithm_input_count=len(inputs),
+                relevant_algorithm_output_count=len(outputs),
+            )
+            .aggregate(
+                max_memory=Max("algorithm_maximum_settable_memory_gb"),
+                gpu_type_choices=ArrayAgg(
+                    "algorithm_selectable_gpu_type_choices", distinct=True
+                ),
+            )
+        )
+
+    @property
+    def selectable_gpu_type_choices(self):
+        choices = [GPUTypeChoices.NO_GPU, GPUTypeChoices.T4]
+        choices.extend(
+            chain.from_iterable(
+                self.job_requirement_properties_from_phases["gpu_type_choices"]
+            )
+        )
+        choices_set = set(choices)
```

Comment:
GitHub messes with the formatting but this might be more straightforward:

```suggestion
        choices_set = {
            GPUTypeChoices.NO_GPU, 
            GPUTypeChoices.T4, 
            *chain.from_iterable(self.job_requirement_properties_from_phases["gpu_type_choices"])
        }
```

Reply:
Yes, that's much cleaner! Thank you